auto_compressor/src/state_saving.rs

```diff
@@ -249,3 +249,44 @@ pub fn write_room_compressor_state(
 
     Ok(())
 }
+
+/// Returns the top [number] rooms with the most uncompressed state
+///
+/// Uncompressed state is measured by number of rows in the state_groups_state
+/// table due to state groups with id's lower than where the compressor has gotten
+/// up to (i.e. the last_compressed value in the state_compressor_progress
+/// table).
+///
+/// # Arguments
+///
+/// * `client`    -   A postgres client used to send the requests to the database
+/// * `number`    -   How many groups to return
+pub fn get_rooms_with_most_rows_to_compress(
+    client: &mut Client,
+    number: i64,
+) -> Result<Option<Vec<(String, i64)>>> {
+    let get_biggest_rooms = r#"
+        SELECT s.room_id, count(*) AS num_rows
+        FROM state_groups_state AS s
+        LEFT JOIN state_compressor_progress AS p 
+            ON s.room_id = p.room_id
+        WHERE s.state_group > p.last_compressed 
+            OR p.last_compressed IS NULL
+        GROUP BY s.room_id
+        ORDER BY num_rows DESC
+        LIMIT $1
```

Comment:
Is this going to be horrible slow when it walks over the full `state_groups_state` table? Or is the idea that `state_compressor_progress` only has a small subset of it?

Reply:
Well yes... It probably will be quite slow, but the idea is that it cues up work for the compressor to do later. If you can think of a better way of doing this then do say! It could be fine with an approximation.

I based the query off of [the wiki](https://github.com/matrix-org/synapse/wiki#what-are-the-biggest-rooms-on-my-server), presuming that such a query couldn't take THAT long if people were being advised to run it. The more the database is compressed, the less time this should take because of the `WHERE state_group > last_compressed`

Reply:
Huh, I'm surprised that's on the wiki. The `state_groups_state` table is almost 5TB on matrix.org, so it would take a *very* long time to walk through it all.

I think we have a couple of options:

1. use `pg_stats` table to get the approximate sizes of the largest rooms, or some other heuristic; or
2. change things around so that we get the first N state groups, compress the rooms that appear there, then look at the next N state groups, etc? It won't do the biggest rooms first but it will eventually get there.


Let me know if you want to talk this through

Reply:
Have opened #72  that uses the 2nd of those two options. This PR might be useful in the future if anyone would rather have the state compressor work this way